_posts/2018-05-21-shipped-and-hosted-software.md

```diff
@@ -0,0 +1,268 @@
+
+Upgrades are important both for new features and for bugfixes, including for
+security bugs. An instance of an application like Gitlab might require an
+immediate upgrade when a security issue is discovered. However, especially if
+the deployment is several versions old, that critical upgrade may carry a great
+deal of risk. Producers of shipped software sometimes provide backported fixes
+for just this purpose, at least for long term support (LTS) or extended support
+release (ESR) versions, but this has a substantial cost for the application
+developers.
+
+Upgrading services like Gitlab or Joomla is made more difficult because there
+is lots of user data that must remain accessible after the upgrade. For major
+upgrades, that often requires some kind of migration as data formats and
+schemas change.  In cases where the upgrade spans several major versions, it
+may be necessary to apply several migrations in order. Tools like Alembic help
+with this by maintaining and applying step-by-step database migrations.
+
+## Taskcluster
+
+Today, Taskcluster is very much a hosted application. There is only one
+"instance" of Taskcluster in the world, at [taskcluster.net](https://taskcluster.net).
+The Taskcluster team is responsible for both development and operation of the
+service, and also works closely with the Firefox build team as a user of the
+service.
+
+We want to make Taskcluster a shipped application. As the descriptions above
+suggest, this is not a simple process. The following sections highlight some
+of the challenges we are facing.
+
+### Releases and Deployment
+
+We currently deploy Taskcluster microservices independently. That is, when we
+make a change to a service like taskcluster-hooks, we deploy an upgrade to that
+service without modifying the other services. We often sequence these changes
+carefully to ensure continued compatibility: we expect only specific
+combinations of services to run together.
+
+This is a far more intricate process than we can expect users to follow.
+Instead, we will ship Taskcluster *releases* comprised of a set of built Docker
+images and a spec file identifying those images and how they should be
+deployed. We will test that this particular combination of versions works well
+together.
+
+Deploying a release involves combining that spec file with some
+deployment-specific configuration and some infrastructure information
+(implemented via [Terraform](http://terraform.io/)) to produce a set of
+[Kubernetes](http://kubernetes.io/) resources for deployment with `kubectl`.
+Kubernetes and Terraform both have limited support for migration from one
+release to another: Terraform will only create or modify changed resources, and
+Kubernetes will perform a phased roll-out of any modified resources.
+
+By the way, all of this build-and-release functionality is implemented in the
+new [taskcluster-installer](https://github.com/taskcluster/taskcluster-installer).
+
+### Service Discovery
```

Comment:
Should we talk about inter-service routing without going back out through ingress here or just ignore it for now?